### PR TITLE
Fix issue with prop inheritance and expose couch base model

### PR DESCRIPTION
**Pull request by ErickFabian**
Fixes issue with prop inheritance not working, add common methods to the `Dolly::Document` class and rename namespace variable

### CODE REVIEW

**Review by javierg — COMMENTED** [review]
When you say to add this to the base I thought it waS on the dolly document.

I dont think we need a `Couch::Base` object, this methods should be implemented in the already existing methods/modules or create new modules, but we will not change the `Dolly::Document` interface. I dont think is a good idea.

File: lib/couch/base.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    def update_attributes!(attributes)
+      attributes.each(&update_attribute)
+      save
+    end
+
+    def update_attributes(attributes)
+
+      doc.merge(attributes)
+      save
+    end
```
Comment (javierg) [inline lib/couch/base.rb]: this methods should go in `DocumentState` module

File: lib/couch/base.rb [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    def base_id
+      self.class.base_id(id)
+    end
+
+    def self.safe_find(id)
+      find(id)
+    rescue Dolly::ResourceNotFound
+      nil
+    end
```
Comment (javierg) [inline lib/couch/base.rb]: this method is part of Query module

File: lib/couch/base.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    alias_method :to_param, :base_id
+    alias_method :write_property, :write_attribute
+
+    def persisted?
```
Comment (javierg) [inline lib/couch/base.rb]: this should be part of doc state module

File: lib/couch/base.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      rev.present?
+    end
+
+    def self.raw_view(name, options = {})
```
Comment (javierg) [inline lib/couch/base.rb]: Lets add this to a view query module

File: lib/couch/base.rb [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      design = options.delete(:design)
+      connection.get("_design/#{design}/_view/#{name}", options)
+    end
+
+    def self.last_record
```
Comment (javierg) [inline lib/couch/base.rb]: there is already a `last` method, we should not need this.

File: lib/couch/base.rb [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      LAST_RECORD
+    end
+
+    def raw_view(name, options = {})
```
Comment (javierg) [inline lib/couch/base.rb]: use a view query module

File: lib/couch/base.rb [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      self.class.raw_view(name, options)
+    end
+
+    def view_value(name, options = {})
```
Comment (javierg) [inline lib/couch/base.rb]: use a view query module

File: lib/couch/base.rb [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    end
+
+    def reload
+      self.doc = self.class.find(id).doc
+    end
+
+    def reload_rev!
+      return unless persisted?
+      rev = connection.get(CGI.escape(id)[:_rev])
+    end
```
Comment (javierg) [inline lib/couch/base.rb]: this 2 not sure if they are a doc state methods or they should be in a service object doing the reload?

File: lib/couch/base.rb [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+    private
+
+    def set_type
```
Comment (javierg) [inline lib/couch/base.rb]: this IMO can be in the namespace module?

File: lib/dolly/view_query.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
+require 'dolly/class_methods_delegation'
+
+module Dolly
+  module ViewQuery
```
Comment (javierg) [inline lib/dolly/view_query.rb:4]: Do you think you can test this methods?

File: lib/dolly/property_manager.rb [hunk 11]
```diff
@@ -43,5 +43,12 @@ def assign_identity_properties(opts = {})
       id_presence = opts[:id] || opts[:_id] || opts['id'] || opts['_id']
       self.id = id_presence if id_presence
     end
+
+    def read_property name
```
Comment (javierg) [inline lib/dolly/property_manager.rb]: where is this method being used?

Comment (ErickFabian) [reply lib/dolly/property_manager.rb]: not used, the change should be done in the apps to use the correct method, i'll remove this

**Review by javierg — APPROVED** [review]
:shipit:

**Pull request merged**